### ISSUE TITLE
Wrong reported errors by compiler

### ISSUE THREAD

**Issue by augustocdias** [redhat-developer/vscode-java#730]
I have some pieces of code that compiles fine in IntelliJ and Gradle but VSCode reports them as errors.
Examples:

1. Error in import

    import sun.security.ssl.ProtocolVersion;

![screenshot 2018-11-30 at 11 46 27](https://user-images.githubusercontent.com/1080789/49284699-a6c47f00-f495-11e8-8149-262914fd44ff.png)

I have to point that I'm not using the JVM in the error description. Here's my workspace's settings:

    {
       "java.home": "/Users/augusto.dias/.sdkman/candidates/java/8.0.181-oracle",
       "java.jdt.ls.vmargs": "-javaagent:/Users/augusto.dias/my-project/.vscode/lombok.jar -Xbootclasspath/a:/Users/augusto.dias/my-project/.vscode/lombok.jar"
    }

2. Wrong exception handling

I have a `RetryTemplate` class from spring and I'm calling `execute` as the following example:

    retryTemplate.execute(getCallback(),getRecovery());

    private RetryCallback<Void, IOException> getCallback() {
        return context -> {
            if (context.getLastThrowable() instanceof IOException) {
                        IOException e = (IOException) context.getLastThrowable();
        // do some other things
                        throw e;
            }
            return null;
        };
    }

    private RecoveryCallback getRecovery() {
        return context -> {
            Throwable lastThrowable = context.getLastThrowable();
        // do some other things
            throw lastThrowable instanceof Exception ? (Exception) lastThrowable : new Exception(lastThrowable);
        }
    }

![screenshot 2018-11-30 at 13 54 27](https://user-images.githubusercontent.com/1080789/49290534-81407100-f4a7-11e8-976a-cb7f62e57ec8.png)

3. And finally: false error in method signature

    assertThat((List) ((Envelope) responseEntity.getBody()).getData(), hasItems(siteDTO));
    assertThat((List) ((Envelope) responseEntity.getBody()).getData(), hasSize(1));

![screenshot 2018-11-30 at 13 57 43](https://user-images.githubusercontent.com/1080789/49290711-fad85f00-f4a7-11e8-9692-0a9c0a932522.png)

##### Environment
- Operating System: macOS Mojave
- JDK version: 1.8
- Visual Studio Code version: 1.29.1
- Java extension version: 0.34.0

**Comment by augustocdias** [comment]
@fbricon sorry for the long time. I didn't have access to my computer during the weekend. 
Here's an example.
If you run `./gradlew clean assemble testClasses`, you'll see that everything compiles, but vscode reports errors in the the test class and also in the app class.

[bug-report.zip](https://github.com/redhat-developer/vscode-java/files/2639337/bug-report.zip)

**Comment by fbricon** [comment]
Can you try opening your project in Eclipse? Looks like an Eclipse compiler issue to me

**Comment by anthonyvdotbe** [comment]
@augustocdias 2. and 3. are due to improper use of generics in the application code (see below). Since the only remaining issue is the incorrect JDK being used, I'd propose to close this issue and, if still reproducible with the latest versions, open a new issue for just that problem.

For 2, the line `private RecoveryCallback getRecovery() {` should be `private RecoveryCallback<Void> getRecovery() {`

For 3, the code in bug-report.zip is:
```
    public void error2() {
        ResponseEntity responseEntity = null;
        assertThat((List) ((Envelope) responseEntity.getBody()).getData(), hasSize(1));
    }
```
but should be:
```
    public void error2() {
        ResponseEntity<Envelope> responseEntity = null;
        assertThat((List<?>) ((Envelope) responseEntity.getBody()).getData(), hasSize(1));
    }
```

**Comment by augustocdias** [comment]
I agree with you that the code could be better but on the other hand I expect the IDE to report as errors what the compiler also reports. I would totally agree with this being shown as warnings, but not errors, since it compiles and runs.

**Comment by Tamaarine** [comment]
This is still an issue. My code with RetryTemplate compiles fine in Intellij but in Vscode it is reported as a syntax error with unhandled exception type Throwable.
